drivers: flash: implement non busy wait polling

mkaranki opened the pull request:
`HAL_OSPI_AutoPolling` is checking is the status match happened in a busy loop. That CPU intensive operation can cause starvation of other tasks especially on flash erase operations which can take seconds to complete.

So, replace the use of `HAL_OSPI_AutoPolling` with
`HAL_OSPI_AutoPolling_IT` when OSPI interrupts are enabled. The match results `HAL_OSPI_StatusMatchCallback` being called.

vai-mikar committed:
drivers: flash: stm32 ospi: implement non busy wait polling

HAL_OSPI_AutoPolling is checking is the status match
happened in a busy loop. That CPU intensive operation
can cause starvation of other tasks especially on flash
erase operations which can take seconds to complete.

So, replace the use of HAL_OSPI_AutoPolling with
HAL_OSPI_AutoPolling_IT. The match results
HAL_OSPI_StatusMatchCallback being called.

Signed-off-by: Miika Karanki <miika.karanki@vaisala.com>

ajarmouni-st reviewed (changes requested):
Hi, the following changes are needed in order to not break STM32H5xx support

```
--- a/drivers/flash/flash_stm32_ospi.h
+++ b/drivers/flash/flash_stm32_ospi.h
@@ -38,7 +38,9 @@
 #define HAL_OSPI_Transmit_DMA HAL_XSPI_Transmit_DMA
 #define HAL_OSPI_Transmit_IT HAL_XSPI_Transmit_IT
 #define HAL_OSPI_AutoPolling HAL_XSPI_AutoPolling
+#define HAL_OSPI_AutoPolling_IT HAL_XSPI_AutoPolling_IT
 #define HAL_OSPI_IRQHandler HAL_XSPI_IRQHandler
+#define HAL_OSPI_Abort HAL_XSPI_Abort
 
 #define HAL_OSPI_ErrorCallback HAL_XSPI_ErrorCallback
 #define HAL_OSPI_CmdCpltCallback HAL_XSPI_CmdCpltCallback
```

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -519,14 +519,40 @@ static bool ospi_address_is_valid(const struct device *dev, off_t addr,
+	/* If interrupt is enabled, no need for busy waiting to check the poll result
+	 * like HAL_OSPI_AutoPolling() does.
+	 */
+	if (irq_is_enabled(DT_IRQN(STM32_OSPI_NODE))) {
```
I don't see cases where irq could be disabled in current driver implementation. Do you have something in mind ?

mkaranki replied:
The branch was left there to handle the polling in driver initialization before the interrupt is enabled.

But I think you're correct that there's no such case where interrupts are disabled but poll is ongoing. So indeed the else branch could be removed for now at least. I'll update the PR.

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -519,14 +519,40 @@ static bool ospi_address_is_valid(const struct device *dev, off_t addr,
+	/* If interrupt is enabled, no need for busy waiting to check the poll result
+	 * like HAL_OSPI_AutoPolling() does.
+	 */
+	if (irq_is_enabled(DT_IRQN(STM32_OSPI_NODE))) {
```
This being said I do see potential cases where irqs could be locked, like for instance if we're to support erase while execution in memory map mode while driver is relocated in RAM. Not supported yet, but under investigation.

In this case, we'd have to fallback on the HAL_OSPI_AutoPolling mode, don't we ?

erwango reviewed (changes requested):
Seem my comments on which I'd like answers.

mkaranki commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -519,14 +519,51 @@ static bool ospi_address_is_valid(const struct device *dev, off_t addr,
  * and nor_rate transfer STR/DTR OSPI_STR_TRANSFER or OSPI_DTR_TRANSFER
  */
-static int stm32_ospi_mem_erased(OSPI_HandleTypeDef *hospi, uint8_t nor_mode, uint8_t nor_rate)
+static int stm32_ospi_mem_erased(struct flash_stm32_ospi_data *dev_data,
```
Yes I think it could. If you see passing quite generic `dev` pointer to these functions is not too generic, I could take a look to refactor these.

mkaranki replied:
`stm32_ospi_mem_ready` is used also on other modes than in the one set in `dev_cfg->data_mode` during the initialization before the mode is changed to the target one. At least I see it's this way. The same applies to `stm32_ospi_write_enable`: parameters can be something else than the final mode/rate. 

`stm32_ospi_mem_erased` could be easily changed to use `dev` pointer since erase operations are not done before changing to the target mode.

vai-mikar committed:
drivers: flash: stm32 ospi refactor stm32_ospi_mem_erased args

Change to use dev pointer, from where the other parameters were
available as well.

Signed-off-by: Miika Karanki <miika.karanki@vaisala.com>

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -519,14 +519,38 @@ static bool ospi_address_is_valid(const struct device *dev, off_t addr,
  */
-static int stm32_ospi_mem_erased(OSPI_HandleTypeDef *hospi, uint8_t nor_mode, uint8_t nor_rate)
+static int stm32_ospi_mem_erased(struct flash_stm32_ospi_data *dev_data,
+		uint8_t nor_mode, uint8_t nor_rate)
```
Non blocking: Indentation.

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -563,21 +587,19 @@ static int stm32_ospi_mem_erased(OSPI_HandleTypeDef *hospi, uint8_t nor_mode, ui
  */
-static int stm32_ospi_mem_ready(OSPI_HandleTypeDef *hospi, uint8_t nor_mode, uint8_t nor_rate)
+static int stm32_ospi_mem_ready(struct flash_stm32_ospi_data *dev_data, uint8_t nor_mode,
+		uint8_t nor_rate)
```
Non blocking: Indentation.

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -613,17 +635,14 @@ static int stm32_ospi_mem_ready(OSPI_HandleTypeDef *hospi, uint8_t nor_mode, uin
 /* Enables writing to the memory sending a Write Enable and wait it is effective */
-static int stm32_ospi_write_enable(OSPI_HandleTypeDef *hospi, uint8_t nor_mode, uint8_t nor_rate)
+static int stm32_ospi_write_enable(struct flash_stm32_ospi_data *dev_data,
+		uint8_t nor_mode, uint8_t nor_rate)
```
Non blocking: Indentation.

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -1109,8 +1123,8 @@ static int flash_stm32_ospi_erase(const struct device *dev, off_t addr,
-			ret = stm32_ospi_mem_ready(&dev_data->hospi,
-						   dev_cfg->data_mode, dev_cfg->data_rate);
+			ret = stm32_ospi_mem_ready(dev_data, dev_cfg->data_mode,
+						dev_cfg->data_rate);
```
Non blocking: Indentation.

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -519,14 +519,38 @@ static bool ospi_address_is_valid(const struct device *dev, off_t addr,
 }
 
+static int stm32_ospi_wait_auto_polling(struct flash_stm32_ospi_data *dev_data,
+		OSPI_AutoPollingTypeDef *s_config, uint32_t timeout_ms)
```
Non blocking: Indentation.

erwango reviewed (approved):
Would be nice to fix indentation, but otherwise LGTM
